### ISSUE TITLE
Mock a specific value or method from a hook, instead of all the hook

### ISSUE THREAD

**Issue by moyolvera** [testing-library/react-hooks-testing-library#350]
Hi. First of all thanks for this awesome library it really help me when dealing with testing hooks.

**The problem:**
A little bg in order to explain myself better. I have a hook that looks like this:

```
import { useState } from 'react';

const awesomeHook = () => {
  /* some other useStates going on on this part */
  const [awesomeValue] = useState(false);

  /* some methods going on on this part */

  return {
    /* bunch of methods and values */,
    awesomeValue
  }
}

export default awesomeHook;
```

And that is being used in a component, lets say <AwesomeComponent />. There i need to be able to mock that awesome value to true and false for different scenarios.
Right now i'm doing this:

```
import * as AwesomeHook from '../hook/awesomeHook';

jest.mock('../hook/awesomeHook');

const mockedAwesomeHook = AwesomeHook as jest.Mocked<typeof AwesomeHook>;
mockedAwesomeHook.default.mockImplementation(() => ({
  /* All the other methods and values */
  awesomeValue: true // i set this true or false depending on scenario
}));
```

The problem is that mocking the entire hook just to change a prop is causing me to have more troubles with other scenarios. Can i mock just that *awesomeValue* instead of all the hook.

I tried to spy on hook but my knowledge on the area is a bit limited, i'll appreciate any help or input on this problem

TL;DR: Mock a specific value or method from a hook, instead of all the hook

**Comment by mpeyper** [comment]
Hi @moyolvera,

Just to clarify, do you want to mock the value that is being returned when the hook is being rendered with `renderHook` (from this library) or when it's being used in a component and being rendered by something else?

In general, I would say that mocking a single value of your hooks response is probably not ideal, unless you are able to mock out what it is setting it (e.g. an API call) and trying to limit the amount of mocking required in your test suite will usually result in more robust and easier to maintain tests, but without a lot more context about the codebase you're working in and the constraints it has, it's very difficult to provide more help with issues like this.

**Comment by henryhobhouse** [comment]
This would be pretty useful if you have a hook returned an object of functions of which some interact with each other. Ideally you want to mock or spy on these interactions and you can't that I currently know of. For example:

```
const useExampleHook = () => {
  const updateById = (id, value) => {
    ... update "thing"
  }

  const toggleParticularValue = (id) => {
    if (...some logic) {
      updateById(id, { particularValue: true })
    } else if (...some other logic) {
      updateById(id, { particularValue: false }
    }
  }

  return {
    updateById,
    toggleParticularValue
  }
}
```

This is likely a poor example but if we take for granted there is use cases for both functions outside of the hook and the pattern makes sense then if I want to test the logic in `toggleParticularValue` I cannot isolate it by mocking `updateById` as far as I know. You can argue that the contract extends to both of them but as they are can be used independently I don't think that argument applies. 

As such we would want the ability to mock individual functions returned from a hook.

**Comment by mpeyper** [comment]
The problem is that there isn't much we can do it intercept that function call to provide a mock implementation as the creation of it is internal to the hook.  Even replacing it in the result object wont work because the second function has closed over the original value and doesn't refer to it by the object reference that is returned.

If you inject the update function into the hook arguments or extract it into another module, then you have a chance:

```js
const updateById = (id, value) => {
    ... update "thing"
}

const useExampleHook = (update = updateById) => {
  const toggleParticularValue = (id) => {
    if (...some logic) {
      update(id, { particularValue: true })
    } else if (...some other logic) {
      update(id, { particularValue: false }
    }
  }

  return {
    updateById: update,
    toggleParticularValue
  }
}
```

or 

```js
import updateById from './updateById'

const useExampleHook = () => {
  const toggleParticularValue = (id) => {
    if (...some logic) {
      updateById(id, { particularValue: true })
    } else if (...some other logic) {
      updateById(id, { particularValue: false }
    }
  }

  return {
    updateById,
    toggleParticularValue
  }
}
```

Often though, it just complicates the code or just isn't practical to do.  With hooks, the latter can often be easier to extract a custom hook that handles the raw operations and the hook you're trying to test handles the abstraction over the top of it:

```js
const useThing() => {
  const updateById = (id, value) => {
    ... update "thing"
  }
  return {updateById
}
```

```js
import useThing from './useThing'

const useExampleHook = () => {
  const { updateById } = useThing()

  const toggleParticularValue = (id) => {
    if (...some logic) {
      updateById(id, { particularValue: true })
    } else if (...some other logic) {
      updateById(id, { particularValue: false }
    }
  }

  return {
    updateById, // perhaps not needed if the caller is able to `useThing` when needed
    toggleParticularValue
  }
}
```

In most cases though, I think you're better off testing the hooks as a black box (i.e. these inputs create these outputs) and not mocking out the dependent calls.  Inject things that create side effects (e.g. stores, apis, etc.) using the context API and mock those if required.  This will create more robust tests that actually test the interface the callers will use without testing the implementation details.
